hw/ip/usbdev/data/usbdev.hjson

```diff
@@ -1286,6 +1286,346 @@
+                Isochronous OUT packets were ignored.
+                '''
+          swaccess: "ro"
+        }
+        {
+          bits: "27:16"
+          name: "endpoints",
+          desc: "Set of endpoints for which this counter is enabled.",
+          swaccess: "rw",
+          hwqe: "true",
+        }
+        {
+          bits: "31",
+          name: "rst",
+          desc: "Write 1 to reset the counter.",
+          swaccess: "wo",
+          hwqe: "true",
+          tags: [// Prevent interaction with the count field.
+                 "excl:CsrNonInitTests:CsrExclWrite"]
+        }
+      ]
+    }
+    { name: "count_drop_rx",
+      desc: "Count of SETUP/OUT packets ignored, dropped or NAKed because the RX FIFO was full.",
+      resval: "0",
+      hwaccess: "hrw",
+      hwext: "true",
+      fields: [
+        {
+          bits: "15:0",
+          name: "count",
+          desc: '''
+                Number of SETUP/OUT packets ignored, dropped or NAKed because the RX FIFO was full.
+                SETUP packets have been ignored, Isochronous OUT packets have been dropped, and
+                non-Isochronous OUT packets have been NAKed.
+                '''
+          swaccess: "ro"
+        }
+        {
+          bits: "27:16"
+          name: "endpoints",
+          desc: "Set of endpoints for which this counter is enabled.",
+          swaccess: "rw",
+          hwqe: "true"
+        }
+        {
+          bits: "31",
+          name: "rst",
+          desc: "Write 1 to reset the counter.",
+          swaccess: "wo",
+          hwqe: "true",
+          tags: [// Prevent interaction with the count field.
+                 "excl:CsrNonInitTests:CsrExclWrite"]
+        }
+      ]
+    }
+    { name: "count_nak_in",
+      desc: "Count of rejected IN transactions; NAK received from host.",
+      resval: "0",
+      hwaccess: "hrw",
+      hwext: "true",
+      fields: [
+        {
+          bits: "15:0",
+          name: "count",
+          desc: '''
+                Number of IN transactions rejected by the host responding with a NAK handshake.
+                This counter may be enabled for a specific set of endpoints in the event that
+                particular pipes are unable to accept more data at the host.
+                '''
+          swaccess: "ro"
+        }
+        {
+          bits: "27:16"
+          name: "endpoints",
+          desc: "Set of endpoints for which this counter is enabled.",
+          swaccess: "rw",
+          hwqe: "true"
+        }
+        {
+          bits: "31",
+          name: "rst",
+          desc: "Write 1 to reset the counter.",
+          swaccess: "wo",
+          hwqe: "true",
+          tags: [// Prevent interaction with the count field.
+                 "excl:CsrNonInitTests:CsrExclWrite"]
+        }
+      ]
+    }
+    { name: "count_nodata_in0",
+      desc: "Count of IN transactions for which no packet data was available.",
```

Comment:
Do these two counters need a common enable and a companion counter of the number of USB frames that they are enabled for? Maybe you can extract that all by software (we have the frame count stamp somewhere already) but since these are mostly counters to make the monitoring more efficient and effective the combo with enable seems better.

Reply:
I'm aiming to leave software to decide the policy of how the counters are used, and build only the mechanism in hardware. Much of the motive for having two 'nodata_in' registers is that we can separate Isochronous IN traffic from non-Isochronous, and the set of endpoints for each counter may be changed at any moment, eg. in response to frame interrupts.

Indeed there's questionable utility to the 'pid_invalid' and 'bitstuff' counters, given that those events could be counted in software.

In situations where heavy traffic/load is _not_ the cause of the events that we're trying to monitor (eg. dropped packets), then it seems reasonable to expect the interrupt servicing to be prompt.